lib/core/output_sink/es_sink.rb

```diff
@@ -9,66 +9,53 @@
 require 'active_support/core_ext/numeric/time'
 require 'app/config'
 require 'core/output_sink/base_sink'
+require 'utility/bulk_queue'
 require 'utility/es_client'
 require 'utility/logger'
+require 'elasticsearch/api'
 
 module Core::OutputSink
   class EsSink < Core::OutputSink::BaseSink
-    def initialize(index_name, request_pipeline, flush_threshold = 50)
+
+    def initialize(index_name, request_pipeline, bulk_queue = Utility::BulkQueue.new)
       super()
       @client = Utility::EsClient.new(App::Config[:elasticsearch])
       @index_name = index_name
       @request_pipeline = request_pipeline
-      @operation_queue = []
-      @flush_threshold = flush_threshold
+      @operation_queue = bulk_queue
     end
 
-    def ingest(document)
-      return if document.blank?
+    private
 
-      @operation_queue << { :index => { :_index => index_name, :_id => document[:id], :data => document } }
-      flush if ready_to_flush?
-    end
+    def do_ingest(id, serialized_document)
```

Comment:
I took this pattern from previous C# experience when a public method of the class is a VERB, and an abstract method it's calling is DO_VERB, e.g:

```C#
class ElasticsearchQueryBase
{
  IEnumerable<ElasticsearchObject> Query(ElasticsearchQuery query) {
    // do something
    return DoQuery(query)
  }
}
```

I'm not sure about the naming, do you have suggestions? I'm not happy with it either at the level that I even thought to switch to composition instead of inheritance and rename methods to something completely different.

Reply:
Now, I think that the naming issue is follows from inheritance pattern. 
I see a several of issues: 
1. Calling `super` in "#initialize" might be not obvious for anyone who wants to create a new sink. 
2. `Base` sink is not expandable because all you can do is change the methods behaviour
3. Some of the methods (`#delete_multiple`) might not be working for other sinks (I don't know if we anticipate other sinks existence)

I see 2 different ways to solve that: 
`Super simple`: Approve and merge because it's not a blocker :)
`Simple`: Get rid of `Base` sink and move its logic into `EsSink` (`Console` and `Output` sinks will have some duplicated code which is okay)
`Complex`: Apply `Strategy` pattern: 

```ruby
class Sink
  def initialize(sink:)
   @sink = sink
  end

  def ingest(document)
   return if document.blank?

   serialized_document = @sink_type.serialize(document)
   @sink.ingest(document['id'], serialized_document)

   increment_ingested_count
   increment_ingested_volume(serialized_document.bytesize)
 end

 def delete(id)
   return if id.nil?
   
   @sink.delete(id)

   increment_deleted_count
 end
end

......
```

When you use it in connector service
```
sink = Sink.new(sink: EsSink.new)
sink.ingest(document)
```

I like the 3 option the most but it's not a blocker (Assuming you will need to rewrite tests)

Reply:
@vidok I see, I've attempted to address this with a bit different naming:

```ruby
sink = Sink.new(EsStrategy.new)
sink.ingest(document)
```

I'm not a big fan of this naming though (EsStrategy? meh EsSink? Not sure either). What do you think?

Reply:
`EsSink` doesn't hurt as much as `EsStrategy` does :), especially if you use named arguments. 

```Ruby
Sink.new(sink: EsSink.new)
```

or 
```Ruby
Sink::Elasticsearch
Sink::Stdout
Sink::Console
```

Reply:
I find it a bit weird about creating a Sink that has another Sink inside, maybe the outer sink should be called differently?

```ruby
Ingester.new(sink: Sink::Elasticsearch.new)
```

Reply:
> I find it a bit weird about creating a Sink that has another Sink inside
agree

> Ingester.new(sink: Sink::Elasticsearch.new)
Yes, why not.